### PR TITLE
Puppet module update

### PR DESCRIPTION
**Pull request by NeatNerdPrime**
* puppet plugin feature update

This is more of a work in progress.
The rationale is that, when you choose to install the puppet plugin, it should handle the installation and configuration of the puppet plugin, Especially the packages and the installation.
this plugin is updated to work with the puppetlabs apt repo site. It is far for complete, but it works well.

### CODE REVIEW

**Comment by andsens** [comment]
Looks nice :-)
Ping me when you got something you think is ready for merging, I'd be happy to review. Also: You might want to add your .project files to your global gitignore file.

**Comment by NeatNerdPrime** [comment]
Will keep you posted ;)

**Comment by NeatNerdPrime** [comment]
Hi @andsens , here's an update, and yeah, should have done that, the gitignore thing. Anyways, awaiting your review.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 1]
```diff
@@ -1,14 +1,32 @@
...
 
+# This is an updated puppet module version, introducing the puppet 4 agent
+# and also enabling you to install puppet modules on the image ready.
+# For now it ASSUMES the following:
+# - Debian Wheezy Jessie install
+# - You want the puppet 4 agent (1.9.+) installed from apt.puppetlabs.com
+#
+# Modules isntalled with this plugin are "--force install"
+
+# TODO: plugin must be able to offer choice of distro package or
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: No TODOs in the code, you can create an issue with a todo list instead. This way people can comment on how to solve it etc.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 2]
```diff
@@ -36,28 +55,82 @@ def run(cls, info):
             raise TaskError(msg)
 
 
-class AddPackages(Task):
-    description = 'Add puppet package'
-    phase = phases.preparation
+class InstallPuppetlabsPC1RepoPackage(Task):
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Aha, now I understand your troubles with the `apt-get update` thing etc.  
All this package does is install the puppetlabs repo, which is a trusted gpg key and a file in `/etc/apt/sources.list.d`. You should instead do that via tasks, check out the apt module in `common/tasks` for how sources are added and keys are trusted (`AddBackports` and `InstallTrustedKeys`).

File: bootstrapvz/plugins/puppet/tasks.py [hunk 3]
```diff
@@ -36,28 +55,82 @@ def run(cls, info):
...
 
+    @classmethod
+    def run(cls, info):
+        puppet_defaults = os.path.join(info.root, 'etc/defaults/puppet')
+        sed_i(puppet_defaults, 'START=no', 'START=yes')        
+        
 class CopyPuppetAssets(Task):
     description = 'Copying puppet assets'
-    phase = phases.system_modification
+    phase = phases.user_modification
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: good catch, that is definitely a user modification :-)

File: bootstrapvz/plugins/puppet/tasks.py [hunk 4]
```diff
@@ -79,14 +152,4 @@ def run(cls, info):
         os.remove(manifest_dst)
 
         hosts_path = os.path.join(info.root, 'etc/hosts')
-        sed_i(hosts_path, '127.0.0.1\s*{hostname}\n?'.format(hostname=hostname), '')
-
-
-class EnableAgent(Task):
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Could you keep the `EnableAgent` task at the bottom? Moving it makes it hard to read the diff.

File: manifests/.gitignore [hunk 5]
```diff
@@ -0,0 +1 @@
+/custom manifests/
```
Comment (andsens) [inline manifests/.gitignore]: not a thing, just don't add that file instead of adding mysterious gitignores :-)

File: bootstrapvz/plugins/puppet/tasks.py [hunk 6]
```diff
@@ -79,14 +152,4 @@ def run(cls, info):
...
-    description = 'Enabling the puppet agent'
-    phase = phases.system_modification
-
-    @classmethod
-    def run(cls, info):
-        puppet_defaults = os.path.join(info.root, 'etc/defaults/puppet')
-        sed_i(puppet_defaults, 'START=no', 'START=yes')
+        sed_i(hosts_path, '127.0.0.1\s*{hostname}\n?'.format(hostname=hostname), '')
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Missing newline at EOF

File: bootstrapvz/plugins/debconf/__init__.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 def validate_manifest(data, validator, error):
     from bootstrapvz.common.tools import log_check_call, rel_path
-    validator(data, rel_path(__file__, 'manifest-schema.yml'))
+    validator(data, rel_path(__file__, 'schema.yaml'))
```
Comment (andsens) [inline bootstrapvz/plugins/debconf/__init__.py]: huh?

**Comment by andsens** [comment]
Oh, also, keep an eye on the test results. [It's pretty bad right now](https://travis-ci.org/andsens/bootstrap-vz/builds/202006401) :-/, you can just run `tox` during development to see if everythings green.

File: .gitignore [hunk 8]
```diff
@@ -14,3 +14,6 @@
 /.tox/
 /build-servers.yml
 /integration.html
+# NeatNerdPrime project files
+.project
+.pydevproject
```
Comment (NeatNerdPrime) [inline .gitignore]: Thanks for pointing that out! Learned something new today, i added my global gitignore. It should be okay

File: bootstrapvz/plugins/puppet/README.rst [hunk 9]
```diff
@@ -1,24 +1,111 @@
...
+want the image to contain the puppet agent software from the puppetlabs repo. 
+You want it to almost contain everything you need to get it up and running 
+This plugin does just that!
+While you're at it, throw in some modules from the forge as well!
+Want to include your own modules? Include them as assets!
+
+This plugin attempts to create more abstraction to the bootstrap-vz build 
+process by creating a robust plugin that enables you to use the most up to date
+Puppetlabs software with your images fresh from the build-oven.
+Before, when using this plugin, you were using the puppet agent software
```
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/README.rst]: My apologies, i've added a short desciption in the changelog and removed these lines from the readme

File: bootstrapvz/plugins/puppet/__init__.py [hunk 10]
```diff
@@ -1,19 +1,22 @@
 import tasks
-
+from distutils.log import info
```
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/__init__.py]: Derp... removed :-)

File: .project [hunk 11]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (NeatNerdPrime) [inline .project]: Thank you again for  pointing that out, should be included in my global gitignore.

File: bootstrapvz/plugins/puppet/tasks.py (hunk 1 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: You're right, will add some issues later on. Again, apologies.

File: bootstrapvz/plugins/puppet/tasks.py (hunk 2 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: Well, i tried to get it *working* and i had some trouble (or fatigue)  implementing this step with the InstallTrustedKeys task. It threw me some crazy gpg errors.. Will create an issue to address this with InstallTrustedKeys. this however also works...

File: bootstrapvz/plugins/puppet/tasks.py (hunk 3 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: You're welcome!

File: bootstrapvz/plugins/puppet/tasks.py (hunk 4 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: Sure thing! Done

File: bootstrapvz/plugins/puppet/tasks.py (hunk 6 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: Fixed with all issues with tox tackled

File: manifests/.gitignore (hunk 5 above)
Comment (NeatNerdPrime) [inline manifests/.gitignore]: I am... Surprised... i have absolutely no idea how that got there. Fixed.

File: bootstrapvz/plugins/debconf/__init__.py (hunk 7 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/debconf/__init__.py]: I think i was really tired when i did my commit... No idea why i did this again... anyways - Ahem - fixed

**Comment by NeatNerdPrime** [comment]
Aside from the  current limitations, this would make a great improvement.

**Comment by NeatNerdPrime** [comment]
@andsens ping! :)

**Comment by NeatNerdPrime** [comment]
@andsens  No good?

**Comment by andsens** [comment]
@NeatNerdPrime, I have just been very busy the last week or so, hence the delay.
For some reason your project files still show up in the PR, have you updated it?  
More importantly, the way you install puppet is not proper, I can't accept the PR if it requires changes to some quite fundamental things in bootstrap-vz, such as the AptUpgrade & AptUpdate tasks. I know that that is the only way to do a standard install for puppet, but you'll have to find a way around it. As mentioned in my review, all that package does is add a GPG key and an apt source, which is already well supported by bootstrap-vz itself. You should be able to add those two things programatically by checking how the ones from the manifest are added.

**Comment by NeatNerdPrime** [comment]
@andsens My apologies if it felt like i was pressuring you, I've been busy as well so i can relate.
Removed the project files, and ensured it again on my fork... also included it in my global gitignire.

Concerning the installation procedure:
At the moment I am making  the changes as requested but run into a particular issue. I'll give an update as soon as I get this resolved.

**Comment by NeatNerdPrime** [comment]
@andsens  there we go, i did the requested changes to the best of my abilities. It now installs the puppet agent by using source/key combo. Hope this is a satisfactory workaround.

**Comment by NeatNerdPrime** [comment]
@andsens Would you be so kind to provide me some feedback? I believe it is ready for merge now.

File: .gitignore [hunk 12]
```diff
@@ -13,4 +13,4 @@
 /.coverage
 /.tox/
 /build-servers.yml
-/integration.html
+/integration.html
```
Comment (andsens) [inline .gitignore]: Missing newline at end of file

File: .project (hunk 11 above)
Comment (andsens) [inline .project]: This should be removed. Still here.

File: .pydevproject [hunk 13]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (andsens) [inline .pydevproject]: This should be removed as well.

File: bootstrapvz/common/task_groups.py [hunk 14]
```diff
@@ -134,6 +134,7 @@ def get_apt_group(manifest):
         group.append(packages.AddTaskselStandardPackages)
     return group
 
+
```
Comment (andsens) [inline bootstrapvz/common/task_groups.py]: unnecessary change.

File: bootstrapvz/common/tasks/packages.py [hunk 15]
```diff
@@ -52,6 +52,7 @@ def install_remote(cls, info, remote_packages):
                                        '--assume-yes'] +
                            map(str, remote_packages),
                            env=env)
+            log_check_call(['chroot', info.root, 'apt-get', 'update'])
```
Comment (andsens) [inline bootstrapvz/common/tasks/packages.py]: There should be no changes to the common tasks for this to work.

File: bootstrapvz/plugins/admin_user/__init__.py [hunk 16]
```diff
@@ -7,8 +7,8 @@ def resolve_tasks(taskset, manifest):
     import logging
     import tasks
     from bootstrapvz.common.tasks import ssh
-
     from bootstrapvz.common.releases import jessie
+
```
Comment (andsens) [inline bootstrapvz/plugins/admin_user/__init__.py]: unnecessary change.

File: bootstrapvz/plugins/file_copy/manifest-schema.yml [hunk 17]
```diff
@@ -18,13 +18,12 @@ properties:
                 group:
                   type: string
               required: [dir]
-              additionalProperties: false
```
Comment (andsens) [inline bootstrapvz/plugins/file_copy/manifest-schema.yml]: Why is this file changed?

File: bootstrapvz/plugins/puppet/tasks.py [hunk 18]
```diff
@@ -1,63 +1,124 @@
+import os
 from bootstrapvz.base import Task
 from bootstrapvz.common import phases
-from bootstrapvz.common.tools import sed_i
-import os
+from bootstrapvz.common.exceptions import TaskError
+from bootstrapvz.common.releases import jessie, wheezy
+from bootstrapvz.common.tools import sed_i, log_check_call
+
+# This is an updated puppet module version, introducing the puppet 4 agent
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: This should be in the docs, not the code.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 19]
```diff
@@ -1,63 +1,124 @@
...
+# For now it ASSUMES the following:
+# - Debian Wheezy Jessie install
+# - You want the puppet 4 agent (1.9.+) installed from apt.puppetlabs.com
+#
+# Modules installed with this plugin are "--force install"
 
 
 class CheckAssetsPath(Task):
-    description = 'Checking whether the assets path exist'
+    description = 'PUPPET: Checking whether the assets path exist'
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Prefixing task descriptions with the module name is non-idiomatic. Please remove or paraphrase to something like "Checking whether the puppet assets path exists"

File: bootstrapvz/plugins/puppet/tasks.py [hunk 20]
```diff
@@ -1,63 +1,124 @@
...
     phase = phases.validation
 
     @classmethod
     def run(cls, info):
         from bootstrapvz.common.exceptions import TaskError
         assets = info.manifest.plugins['puppet']['assets']
         if not os.path.exists(assets):
-            msg = 'The assets directory {assets} does not exist.'.format(assets=assets)
+            msg = 'PUPPET: The assets directory {assets} does not exist.'.format(assets=assets)
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Same here with the "PUPPET:" thing, and further down.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 21]
```diff
@@ -1,63 +1,124 @@
...
 
     @classmethod
     def run(cls, info):
-        info.packages.add('puppet')
+        if info.manifest.release == jessie:
+            info.source_lists.add('puppetlabs', 'deb http://apt.puppetlabs.com jessie PC1')
+        elif info.manifest.release == wheezy:
+            info.source_lists.add('puppetlabs', 'deb http://apt.puppetlabs.com wheezy PC1')
+        else:
+            msg = 'PUPPET: The chosen Debian release \"{system.release}\"Is not (yet) supported by Puppetlabs.com'
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: You should add the release check in either the `validate_manifest()` function or have a separate task for this in the validation phase.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 22]
```diff
@@ -1,63 +1,124 @@
...
-class CopyPuppetAssets(Task):
-    description = 'Copying puppet assets'
+class InstallPuppetlabsPC1ReleaseKey(Task):
+    description = 'PUPPET: Install puppetlabs PC1 Release key into the keyring'
+    phase = phases.package_installation
+
+    @classmethod
+    def run(cls, info):
+        key_file_name = 'DEB-GPG-KEY-puppet'
+        if (info.manifest.release == jessie or info.manifest.release == wheezy):
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: Unnecessary check, you already did that.

File: bootstrapvz/plugins/file_copy/manifest-schema.yml (hunk 17 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/file_copy/manifest-schema.yml]: TL;DR: Dont drink and code

I believe this is a residue from something else I've been working on, sorry about that, fixed

File: bootstrapvz/plugins/puppet/tasks.py (hunk 21 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: I've chosen to split it up in a separate task. hope this meets your request

**Comment by NeatNerdPrime** [comment]
@andsens First of all, thank you for your patience, it has been busy for me as well. Without further ado here is my newest commit

**Comment by NeatNerdPrime** [comment]
@andsens Would you be so kind to provide me some feedback? I believe it is ready for merge now.

**Comment by NeatNerdPrime** [comment]
@andsens Happy Eastern! (in case you celebrate it). Would you be so kind to review my changes and merge them if they are satisfactory? I have performed all the changes requested and i believe it is ready to merge now.

**Comment by NeatNerdPrime** [comment]
@andsens Don't wish to be rude but, should I take your absence of a response as "not interested/bad PR"?

**Comment by andsens** [comment]
Hi there, I just came back from vacation, hence the delayed response.  
`bootstrapvz/common/task_groups.py` and `bootstrapvz/plugins/file_copy/manifest-schema.yml` still have unrelated changes that should be removed, you should also squash those 31 commits into a single commit with a nice message, after that I think we're good to go.

**Comment by NeatNerdPrime** [comment]
@andsens Ha, I hope your vacation was enjoyable with lots of good time. 
I've tried to squash all the commits into a neat one. However i do not seem to be able to remove the files `bootstrapvz/common/task_groups.py` and 'bootstrapvz/plugins/file_copy/manifest-schema.yml' even when doing 'git rm <files> ; git commit --amend'. 

I admit i'm not a git guru, learned a lot with doing this work, but i cannot get my head around that one. Any (last) recommendation on how to fix that?

Again thank you for your patience and time to review.

**Comment by andsens** [comment]
Run `git checkout 4fbfd795 -- bootstrapvz/common/task_groups.py bootstrapvz/plugins/file_copy/manifest-schema.yml`. This will get you the original files, then you can just `commit --amend` them. git will then see it as a no-op and remove any mention of the files from the commit.

**Comment by NeatNerdPrime** [comment]
@andsens This should do it!

File: bootstrapvz/plugins/puppet/tasks.py [hunk 23]
```diff
@@ -36,28 +52,75 @@ def run(cls, info):
...
+            info.source_lists.add('puppetlabs', 'deb http://apt.puppetlabs.com wheezy PC1')
+
+
+class InstallPuppetlabsPC1ReleaseKey(Task):
+    description = 'Install puppetlabs PC1 Release key into the keyring'
+    phase = phases.package_installation
+
+    @classmethod
+    def run(cls, info):
+        key_file_name = 'DEB-GPG-KEY-puppet'
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: I was just wondering, why not just do like [`InstallTrustedKeys`](https://github.com/andsens/bootstrap-vz/blob/7af9cd9fb85d845884b2bd031fa694cc3c101d81/bootstrapvz/common/tasks/apt.py#L111) does?
i.e. `git add` the key to `plugins/puppet/assets/gpg-key` and then install it by copying:
```
from shutil import copy
key_path = os.path.join(ASSETS_DIR, 'DEB-GPG-KEY-puppet')
destination = os.path.join(info.root, 'etc/apt/trusted.gpg.d/DEB-GPG-KEY-puppet')
copy(key_path, destination)
```

where `ASSETS_DIR` is define in the top like in the [expand_root plugin](https://github.com/andsens/bootstrap-vz/blob/7af9cd9fb85d845884b2bd031fa694cc3c101d81/bootstrapvz/plugins/expand_root/tasks.py#L11). Like so: `ASSETS_DIR = rel_path(__file__, 'assets')`

This way you avoid running a load of commands and you don't install `ca-certificates`, which the user didn't specify as a package really.

Sorry for only seeing this now :-/

**Comment by andsens** [comment]
Yup, looks good :-)
I added a little comment on how to improve the code, I can merge without it, but there is the issue that you are actually installing `ca-certificates` to the image. The user has never specified that package, so it might be a little wrong to just install it for them.

**Comment by NeatNerdPrime** [comment]
@andsens You're right, it is not needed to install the package this roundabout way. I've updated my PR with the changes that support your request. It now won't install the 'ca-certificates' package anymore, keyrings get injected directly. I believe i did a pretty good job here. You concur?

File: bootstrapvz/plugins/puppet/tasks.py (hunk 23 above)
Comment (NeatNerdPrime) [inline bootstrapvz/plugins/puppet/tasks.py]: @andsens  At the time I didnt manage to get it working that way, so i tried the other approach by fetching the key from the internet itself. the reason why the package ca-certificates was added is to ensure the wget command would succeed and still verify the SSL connection to the apt.puppetlabs.com repo.

**Comment by andsens** [comment]
Quick update for all PRs: I've seen them and will get to them. I'm a little pre-occupied at the moment, so it might be two-three weeks before I answer. (and thanks for the PR btw :-))

**Comment by NeatNerdPrime** [comment]
@andsens Hi, it's been a while, you'll notice that I've added a bit more functionality to the module. Would you please give me your review and - if approved -  merge it? Thanks in advance.

File: bootstrapvz/plugins/puppet/example/jessie-virtio-with-puppet.yaml [hunk 24]
```diff
@@ -0,0 +1,45 @@
+---
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/example/jessie-virtio-with-puppet.yaml]: You should put that in `manifests/examples/kvm/jessie-puppet.yml`. When building the python package this folder wont be included.

File: bootstrapvz/plugins/puppet/tasks.py [hunk 25]
```diff
@@ -36,48 +56,93 @@ def run(cls, info):
...
+
+    @classmethod
+    def run(cls, info):
+        for module in info.manifest.plugins['puppet']['install_modules']:
+            command = ['chroot', info.root, '/opt/puppetlabs/bin/puppet', 'module', 'install', '--force']
+            if (len(module) == 1):
+                module_name = module[0]
+                command.append(str(module_name))
+            if (len(module) == 2):
+                module_name = module[0]
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: You could just deconstruct this:
```
[module_name, module_version] = module
```
:-)

File: bootstrapvz/plugins/puppet/tasks.py [hunk 26]
```diff
@@ -36,48 +56,93 @@ def run(cls, info):
...
     def run(cls, info):
         from bootstrapvz.common.tools import copy_tree
-        copy_tree(info.manifest.plugins['puppet']['assets'], os.path.join(info.root, 'etc/puppet'))
+        copy_tree(info.manifest.plugins['puppet']['assets'], os.path.join(info.root, 'etc/puppetlabs/'))
 
 
 class ApplyPuppetManifest(Task):
-    description = 'Applying puppet manifest'
-    phase = phases.user_modification
+    description = 'Applying puppet manifest...'
```
Comment (andsens) [inline bootstrapvz/plugins/puppet/tasks.py]: No ellipsis (...), they would apply to all task descriptions really :-)

**Comment by NeatNerdPrime** [comment]
@andsens  this should do it.

**Comment by andsens** [comment]
Yup, looks good 👍 , merging.
Nice job @NeatNerdPrime!
